### PR TITLE
[Feature: Forum] Create Socket Server script and Push threads/posts to users

### LINKED ISSUE #4328: Use WebSockets to live fetch updates on Forum
**What problem are you trying to solve with Submitty**
The current Forum fetches data from the server upon refresh. This causes an unnecessary load on the server, and also restricts the users from getting automatic fresh updates. 

**Describe the way you'd like to solve this problem**
I am adding a WebSocket to the Forum (sitewide available, actually) which can be used to facilitate a bi-directional transfer of information from the server, keeping a live connection. 

**Describe any potential alternatives you'd tried to solve the problem**
This could probably be solved with interval based fetching of data from the server.

**Additional context**
This feature request has been put in the form of a proposal at - https://docs.google.com/document/d/1phqFPUhyv6-1IC-ncrldKcRWEcFxjd2dyMlf9UgUSis/edit#heading=h.5uoc4mfz7mn4

### PR DESCRIPTION
**Pull request by xprilion**
### Please check if the PR fulfills these requirements:

* [x] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Partially Closes #4328 

### What is the new behavior?
Addition of Web Socket to the system, which is connected to by the discussion Forum threads listing page. 

### How to test: 

1. `vagrant ssh`
2. `php site/app/controllers/SocketController.php`
3. Open the Discussion Forum
4. Enable the Chrome Console
5. Open another instance of Forum on another browser/incognito
6. Make replies to posts, create threads to observe Forum updates

### CODE REVIEW

**Comment by MasterOdin** [comment]
I'd like to see as part of this PoC that it includes the mechanisms by which we'd actually push content to the connected clients.

File: migration/migrator/data/course_tables.sql [hunk 1]
```diff
@@ -644,6 +644,18 @@ CREATE TABLE "viewed_responses" (
     CONSTRAINT viewed_responses_pkey PRIMARY KEY ("thread_id", "user_id")
 );
 
+CREATE TABLE socket_connections
+(
+  connection_id SERIAL NOT NULL,
+  user_id       VARCHAR(500),
```
Comment (zjxiaohan) [inline migration/migrator/data/course_tables.sql]: `user_id`, as defined elsewhere, should be `character varying(255)`

I am not sure how socket works, so sorry for the dumb question: where is this table used? I cannot find code for inserting records into this table.

Comment (xprilion) [reply migration/migrator/data/course_tables.sql]: @zjxiaohan thanks for pointing that out! I'll quickly update that, was hoping to do so in future updates to this PR...also there's no code for that current, I just wanted to see if these changes were failing the build so didn't add a WIP, will do so!

**Comment by xprilion** [comment]
@MasterOdin I was showing the same to @andrewaikens87 the other day on the meeting, while working on getting post replies propagated currently, as a first! Would you like to join in on the next meeting, for a demo? :D

**Comment by andrewaikens87** [comment]
I agree with @MasterOdin. In terms of this PR, we should have it be a PoC so that it can be tested and reviewed. This PR was made to get any initial feedback as @xprilion moves forward. This shouldn't be merged as is until the feature is implemented. This should probably be renamed to WIP for the time being.

File: site/app/models/NotificationFactory.php [hunk 2]
```diff
@@ -1,19 +1,16 @@
...
 use app\libraries\Core;
-use app\models\Email;
-use app\models\Notification;
-use app\models\User;
 use LogicException;
 
 /**
  * A factory class that will handle all notification events and send notifications and emails accordingly
  * Class NotificationFactory
  * @package app
```
Comment (andrewaikens87) [inline site/app/models/NotificationFactory.php]: Also change comment if namespace is changing.

File: site/app/models/NotificationFactory.php [hunk 3]
```diff
@@ -1,19 +1,16 @@
...
  */
-class NotificationFactory {
+class NotificationFactory extends AbstractModel {
```
Comment (andrewaikens87) [inline site/app/models/NotificationFactory.php]: Will loop back in on this. Not sure if this change is required.

File: site/app/templates/forum/ShowForumThreads.twig [hunk 4]
```diff
@@ -3,6 +3,7 @@
 <script>
 
     $( document ).ready(function(){
+
```
Comment (andrewaikens87) [inline site/app/templates/forum/ShowForumThreads.twig]: Please remove newline.

File: site/app/templates/forum/ThreadPostForm.twig [hunk 5]
```diff
@@ -42,38 +47,38 @@
         </label>
         <script>
             $(document).ready(function () {
-                        flatpickr(".date_picker", {
-                            plugins: [ShortcutButtonsPlugin(
+                flatpickr(".date_picker", {
```
Comment (andrewaikens87) [inline site/app/templates/forum/ThreadPostForm.twig]: Is this change required? Seems just like a spacing change.

File: site/app/templates/forum/ThreadPostForm.twig [hunk 6]
```diff
@@ -154,6 +159,7 @@
                     $(".upload_attachment_box").each(function() {
                         this.addEventListener("click", clicked_on_box, false);
                     });
+
```
Comment (andrewaikens87) [inline site/app/templates/forum/ThreadPostForm.twig]: Please remove newline.

File: site/app/views/forum/ForumThreadView.php [hunk 7]
```diff
@@ -550,6 +553,26 @@ public function generatePostList($currentThread, $posts, $unviewed_posts, $curre
         return $return;
     }
 
+    function renderSingleThreadListItem($threads){
```
Comment (andrewaikens87) [inline site/app/views/forum/ForumThreadView.php]: Please add comments.

File: site/app/views/forum/ForumThreadView.php [hunk 8]
```diff
@@ -550,6 +553,26 @@ public function generatePostList($currentThread, $posts, $unviewed_posts, $curre
...
+        $currentCategoriesIds = $this->core->getQueries()->getCategoriesIdForThread($currentThread);
+        $activeThreadAnnouncement = false;
+        $activeThreadTitle = "";
+        $activeThread = array();
+        $displayThreadContent = $this->displayThreadList([$thread], false, $activeThreadAnnouncement, $activeThreadTitle, $activeThread, $currentThread, $currentCategoriesIds, false);
+        $html = $this->core->getOutput()->renderTwigTemplate("forum/displayThreadList.twig", $displayThreadContent);
+        $this->core->getOutput()->renderJsonSuccess($html);
+    }
+
+    function renderPost($thread_id, $post, $post_box_id){
```
Comment (andrewaikens87) [inline site/app/views/forum/ForumThreadView.php]: Please add comments.

File: site/app/views/forum/ForumThreadView.php (hunk 7 above)
Comment (andrewaikens87) [reply site/app/views/forum/ForumThreadView.php]: Also include visibility modifier. (even though default is public)

File: site/app/views/forum/ForumThreadView.php (hunk 8 above)
Comment (andrewaikens87) [reply site/app/views/forum/ForumThreadView.php]: Also include visibility modifier. (even though default is public)

File: site/app/views/forum/ForumThreadView.php [hunk 9]
```diff
@@ -766,7 +789,7 @@ public function createPost($thread_id, $post, $unviewed_posts, $function_date, $
         $thread_dir = FileUtils::joinPaths(FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "forum_attachments"), $thread_id);
 
         $date = DateUtils::parseDateTime($post["timestamp"], $this->core->getConfig()->getTimezone());
-        if (!is_null($post["edit_timestamp"])) {
+        if (isset($post["edit_timestamp"]) && !is_null($post["edit_timestamp"])) {
```
Comment (andrewaikens87) [inline site/app/views/forum/ForumThreadView.php]: Can `!empty()` be used here?

File: site/app/views/forum/ForumThreadView.php [hunk 10]
```diff
@@ -943,8 +966,10 @@ public function createPost($thread_id, $post, $unviewed_posts, $function_date, $
             "form_post_url" => $this->core->buildCourseUrl(['forum', 'posts', 'new']),
             "post_box_id" => $post_box_id,
             "thread_id" => $thread_id,
-            "parent_id" => $post_id,
-            "render_markdown" => $markdown
+            "parent_id" => $post["parent_id"],
+            "render_markdown" => $markdown,
+            "csrf_token" => $this->core->getCsrfToken(),
+            "post_content_limit" => ForumUtils::FORUM_CHAR_POST_LIMIT
```
Comment (andrewaikens87) [inline site/app/views/forum/ForumThreadView.php]: Was this change missed in a previous PR? I'm just trying to understand why it is added here.

File: site/app/templates/forum/ThreadPostForm.twig (hunk 5 above)
Comment (xprilion) [reply site/app/templates/forum/ThreadPostForm.twig]: Was just making the code in proper alignment..

File: site/app/views/forum/ForumThreadView.php (hunk 10 above)
Comment (xprilion) [reply site/app/views/forum/ForumThreadView.php]: This was previously not available in the way CreatePost worked. It was being called within other twig files, which were getting the constant from their View function. CreatePost being called directly here requires the passing of this constant to be able to use it.

File: site/app/views/forum/ForumThreadView.php (hunk 9 above)
Comment (xprilion) [reply site/app/views/forum/ForumThreadView.php]: No, as per what I read from https://www.virendrachandak.com/techtalk/php-isset-vs-empty-vs-is_null/

`isset( )` is probably the correct one here

File: site/app/controllers/SocketController.php [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<?php
+    namespace app\controllers;
+
+    use Ratchet\Server\IoServer;
+    use Ratchet\Http\HttpServer;
+    use Ratchet\WebSocket\WsServer;
+    use app\libraries\Socket;
+
+    require dirname(__DIR__) . '/../vendor/autoload.php';
```
Comment (zjxiaohan) [inline site/app/controllers/SocketController.php]: require_once may be better. The require_once statement is identical to require except PHP will check if the file has already been included, and if so, not include (require) it again.

https://www.php.net/manual/en/function.require-once.php

File: site/app/views/forum/ForumThreadView.php (hunk 9 above)
Comment (andrewaikens87) [reply site/app/views/forum/ForumThreadView.php]: This whole conditional can be replaced with `if(!empty())` no need to have isset and !is_null. Empty doesn't trigger an error if the variable doesn't exist.

**Comment by andrewaikens87** [comment]
image

If a course has an announcement a normal thread can be populated above it. Normal threads should always fall under announcements.

**Comment by andrewaikens87** [comment]
Can we catch this exception from being thrown in the JS if the socket server is not active? 
image

**Comment by andrewaikens87** [comment]
image

Also ensure that we cap the offset to a max right shift. This is done in the php code and should be replicated when we place the replies via the JS.

**Comment by xprilion** [comment]
Hey @andrewaikens87 I was unable to catch the WS not connected error :( 

I did solve the other two issues!

File: site/app/controllers/SocketController.php (hunk 11 above)
Comment (xprilion) [reply site/app/controllers/SocketController.php]: I'm not sure if that is what we need here, I guess I just copied that line from another place within Submitty :o

Comment (zjxiaohan) [reply site/app/controllers/SocketController.php]: Where is it? `index.php` is using `require_once` though...

**Comment by andrewaikens87** [comment]
As @MasterOdin mentioned, I fixed the service file. Should work now. Gonna wait to see if travis passes.

File: site/socket/SocketListener.php [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    require_once(__DIR__.'/../vendor/autoload.php');
+
+    $server = IoServer::factory(
+        new HttpServer(
+                new WsServer(
+                    new Socket()
+                )
+        ),
+        8080
+    );
```
Comment (MasterOdin) [inline site/socket/SocketListener.php]: Can we limit this to only listening on `127.0.0.1` (after doing the aforementioned apache stuff) and not have this port be public, or is this something we'd have to do via iptables?

**Comment by pratiklodha** [comment]
Hi! This is one of the listed projects for GSoC 2020. From discussion threads, it seems that this feature is implemented in PHP. Actually, I am too interested in this project. I have developed a web-based chat application (real-time message display without page reloading) using web sockets but in Nodejs.

I have zero knowledge of PHP. Is it possible to contribute using JS (Nodejs)?

**Comment by pratiklodha** [comment]
Hi! This is one of the listed projects for GSoC 2020. From discussion threads, it seems that this feature is implemented in PHP. Actually, I am too interested in this project. I have developed a web-based chat application (real-time message display without page reloading) using web sockets but in Nodejs.

I have zero knowledge of PHP. Is it possible to contribute using JS (Nodejs)?

**Comment by pratiklodha** [comment]
Hi! This is one of the listed projects for GSoC 2020. From discussion threads, it seems that this feature is implemented in PHP. Actually, I am too interested in this project. I have developed a web-based chat application (real-time message display without page reloading) using web sockets but in Nodejs.

I have zero knowledge of PHP. Is it possible to contribute using JS (Nodejs)?

**Comment by MasterOdin** [comment]
Closing this with the websocket handled in #5205 and the forum stuff being (re-)done in new follow-up PRs. Thanks for getting us the right path with this @xprilion!

**Comment by xprilion** [comment]
That's great news @MasterOdin! Looking forward to seeing the sockets in action on Submitty 😁

**Pull request closed without merging**